incubator/virtualcluster/pkg/syncer/conversion/mutate.go

```diff
@@ -0,0 +1,251 @@
+		if !v1helper.IsServiceIPSet(service) {
+			continue
+		}
+		serviceName := service.Name
+
+		// We always want to add environment variabled for master services
+		// from the master service namespace, even if enableServiceLinks is false.
+		// We also add environment variables for other services in the same
+		// namespace, if enableServiceLinks is true.
+		if service.Namespace == masterServiceNamespace && masterServices.Has(serviceName) {
+			if _, exists := serviceMap[serviceName]; !exists {
+				serviceMap[serviceName] = service
+			}
+		} else if service.Namespace == ns && enableServiceLinks {
+			serviceMap[serviceName] = service
+		}
+	}
+
+	var mappedServices []*v1.Service
+	for key := range serviceMap {
+		mappedServices = append(mappedServices, serviceMap[key])
+	}
+
+	for _, e := range envvars.FromServices(mappedServices) {
+		m[e.Name] = e.Value
+	}
+	return m
+}
+
+func (p *podMutator) mutateDNSConfig(vPod *v1.Pod, nameServer string) {
+	dnsPolicy := p.pPod.Spec.DNSPolicy
+
+	switch dnsPolicy {
+	case v1.DNSNone:
+		return
+	case v1.DNSClusterFirstWithHostNet:
+		p.mutateClusterFirstDNS(vPod, nameServer)
+		return
+	case v1.DNSClusterFirst:
+		if !p.pPod.Spec.HostNetwork {
+			p.mutateClusterFirstDNS(vPod, nameServer)
+			return
+		}
+		// Fallback to DNSDefault for pod on hostnetwork.
+		fallthrough
+	case v1.DNSDefault:
+		// FIXME(zhuangqh): allow host dns or not.
+		p.pPod.Spec.DNSPolicy = v1.DNSNone
+		return
+	}
+}
+
+func (p *podMutator) mutateClusterFirstDNS(vPod *v1.Pod, nameServer string) {
+	dnsConfig := &v1.PodDNSConfig{}
+	existingDNSConfig := p.pPod.Spec.DNSConfig
+
+	// For a pod with DNSClusterFirst policy, the cluster DNS server is
+	// the only nameserver configured for the pod. The cluster DNS server
+	// itself will forward queries to other nameservers that is configured
+	// to use, in case the cluster DNS server cannot resolve the DNS query
+	// itself.
+	// FIXME(zhuangqh): tenant configure more dns server.
+	dnsConfig.Nameservers = []string{nameServer}
+
+	if p.vc.ClusterDomain != "" {
+		nsSvcDomain := fmt.Sprintf("%s.svc.%s", vPod.Namespace, p.vc.ClusterDomain)
+		svcDomain := fmt.Sprintf("svc.%s", p.vc.ClusterDomain)
+		dnsConfig.Searches = []string{nsSvcDomain, svcDomain, p.vc.ClusterDomain}
+	}
+
+	if existingDNSConfig != nil {
+		dnsConfig.Nameservers = omitDuplicates(append(existingDNSConfig.Nameservers, dnsConfig.Nameservers...))
+		dnsConfig.Searches = omitDuplicates(append(existingDNSConfig.Searches, dnsConfig.Searches...))
+	}
+
+	p.pPod.Spec.DNSPolicy = v1.DNSNone
+	p.pPod.Spec.DNSConfig = dnsConfig
+}
+
+func omitDuplicates(strs []string) []string {
+	uniqueStrs := make(map[string]bool)
+
+	var ret []string
+	for _, str := range strs {
+		if !uniqueStrs[str] {
+			ret = append(ret, str)
+			uniqueStrs[str] = true
+		}
+	}
+	return ret
+}
+
+type ServiceMutateInterface interface {
+	Mutate()
+}
+
+type serviceMutator struct {
+	vc       *v1alpha1.VirtualclusterSpec
```

Comment:
Does service mutator need VC at all?

Reply:
nope. But maybe it does later. This is a unify framework for mutator